### PR TITLE
fix: track multiple pending WHOAREYOU challenges during handshake

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR Description

A WHOAREYOU issued for a new ordinary packet nonce used to overwrite any prior pending challenge on the same session. When the initiator's handshake — signed against the earlier challenge — finally arrived, `HandshakeMessagePacketHandler` validated it against the replacement and dropped the session.

`NodeSession` now keeps a bounded (LRU, max 5) set of pending WHOAREYOU challenges keyed by the ordinary packet nonce that caused each one. `HandshakeMessagePacketHandler` iterates the pending challenges and accepts the one whose ID signature verifies, then clears the set. The set is also cleared on `resetHandshakeState()`.

Behavior summary:
- Same ordinary nonce seen again → resend the existing WHOAREYOU for that nonce (unchanged from #219).
- Different ordinary nonce while waiting for a handshake → queue a new WHOAREYOU without invalidating earlier pending challenges.
- Handshake arrives → match against any unexpired pending challenge.
- Successful authentication or handshake-state reset → clear pending challenges.
- LRU bound keeps the cache memory-safe under load.

### Tests
- Regression test `HandshakeHandlersTest.responderAcceptsHandshakeForEarlierPendingWhoAreYou` reproduces the interleaving in the issue: `n1 → c1`, handshake-for-c1 held, `n2 → c2`, then handshake-for-c1 delivered — responder must authenticate.
- New `NodeSessionTest` cases covering multi-challenge storage, LRU eviction, `clearPendingWhoAreYouChallenges`, and `resetHandshakeState` clearing.
- Updated `UnauthorizedMessagePacketHandlerTest` for the new `hasPendingWhoAreYouForNonce` / `resendOutgoingWhoAreYouFor(nonce)` API.

## Fixed Issue(s)
fixes #228

---

> [!NOTE]
> **Medium Risk**
> Changes discv5 handshake state tracking and signature validation across multiple pending WHOAREYOU challenges, which can affect session authentication behavior under packet reordering/retransmission. Logic is bounded and covered by new regression/unit tests, but it touches core handshake flow.
> 
> **Overview**
> Fixes a handshake edge case where issuing a new `WHOAREYOU` would overwrite the prior pending challenge, causing delayed handshakes to fail authentication.
> 
> `NodeSession` now retains a bounded (LRU, max `5`) map of pending `WHOAREYOU` challenges keyed by ordinary packet nonce, adds nonce-specific resend APIs, and clears all pending challenges on successful authentication or handshake-state reset.
> 
> `HandshakeMessagePacketHandler` now tries *all* pending challenges to find one whose ID signature verifies before deriving session keys, and `UnauthorizedMessagePacketHandler` resends the existing challenge only when the incoming nonce matches a queued one; tests were updated and expanded with a regression for the interleaving scenario plus LRU/clearing coverage.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5e1f75f6455facd61e5a9300e7bf7608b042d253. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/main/java/org/ethereum/beacon/discovery/schema/NodeSession.java [hunk 1]
```diff
@@ -68,8 +79,21 @@ public class NodeSession {
...
+  // been issued for a different ordinary packet. Keeping the set bounded prevents an attacker (or
+  // a busy peer) from forcing unbounded growth.
+  private final Map<Bytes12, PendingWhoAreYou> pendingWhoAreYou =
+      Collections.synchronizedMap(
+          new LinkedHashMap<>(MAX_PENDING_WHOAREYOU + 1, 1.0f, false) {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<Bytes12, PendingWhoAreYou> eldest) {
+              return size() > MAX_PENDING_WHOAREYOU;
+            }
+          });
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply src/main/java/org/ethereum/beacon/discovery/schema/NodeSession.java:95]: @usmansaleem I think this comment is correct.

File: src/test/java/org/ethereum/beacon/discovery/HandshakeHandlersTest.java [hunk 2]
```diff
@@ -243,6 +243,139 @@ public void authHandlerWithMessageRoundTripTest() throws Exception {
...
+                nodeRecord2, signer2, NodeRecordListener.NOOP, NewAddressHandler.NOOP),
+            new LivenessChecker(clock));
+
+    LinkedBlockingQueue<RawPacket> outgoing1Packets = new LinkedBlockingQueue<>();
+    final Consumer<NetworkParcel> outgoingMessages1to2 =
+        parcel -> outgoing1Packets.add(parcel.getPacket());
+    final ExpirationSchedulerFactory expirationSchedulerFactory =
+        new ExpirationSchedulerFactory(Executors.newSingleThreadScheduledExecutor());
+    final ExpirationScheduler<Bytes> requestExpirationScheduler =
+        expirationSchedulerFactory.create(60, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (lucassaldanha) [reply src/test/java/org/ethereum/beacon/discovery/HandshakeHandlersTest.java:277]: This one we can probably ignore.

File: src/main/java/org/ethereum/beacon/discovery/schema/NodeSession.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply src/main/java/org/ethereum/beacon/discovery/schema/NodeSession.java:95]: Good catch — fixed in 8954e59. Flipped `accessOrder` to `true` and added `pendingWhoAreYou_shouldEvictLeastRecentlyAccessedEntry` which fails on the FIFO behavior and passes after the change.

File: src/test/java/org/ethereum/beacon/discovery/HandshakeHandlersTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply src/test/java/org/ethereum/beacon/discovery/HandshakeHandlersTest.java:277]: Leaving this as-is — it mirrors the existing `authHandlerWithMessageRoundTripTest` in the same file, which also doesn't shut down its scheduler. Happy to address it separately if we want to tidy up both.

**Review by lucassaldanha — APPROVED** [review]
